Issue/Enhancement-47 More Formats

henry739 opened the pull request:
This PR fixes #47 (and indirectly fixes #428) and ends up with the following:
## Set-Centric Outputs
- Standard.json
- Modern.json
- Legacy.json
- Vintage.json

## Card-Centric Outputs
- StandardCards.json
- ModernCards.json
- LegacyCards.json
- VintageCards.json
- PauperCards.json

## Other
In addition, the old implementations for retrieving modern sets (from Gamepedia) and standard have been replaced with a more general approach that uses data we already have.

henry739 committed:
add new format module and find_formats_for_sets

Given a path to an AllSets.json file and a list of valid formats, computes and returns a dictionary mapping set code to legal sets.

Note: Does not currently work for pauper or penny dreadful.

henry739 committed:
add build_format_subset and improve variable naming

henry739 committed:
rename to build_format_map and make return value format-centric

henry739 committed:
implement filtering based on set type

henry739 committed:
add more outputs and make build_format_map more flexible

henry739 committed:
add unit test for build_format_map

henry739 committed:
add set output bundle for future standard

henry739 committed:
add future output constant

henry739 committed:
fix range of formatting and style isses

ZeldaZach commented on mtgjson4/format.py:
```diff
@@ -0,0 +1,50 @@
+"""
+from typing import List, Dict, Any
+
+FORMATS = (
```
Define as set via brackets, since we use it that way later in the code

ZeldaZach replied:
Also, define the type `Set[str]`

ZeldaZach commented on mtgjson4/format.py:
```diff
@@ -0,0 +1,50 @@
+    a list of legal formats.
+
+    :param all_sets: AllSets content
+    :type all_sets: dict
```
I don't expect types in the formats, just params :)

ZeldaZach commented on mtgjson4/format.py:
```diff
@@ -0,0 +1,50 @@
+    formats: Dict[str, List[str]] = {fmt: [] for fmt in FORMATS}
+
+    for code, data in all_sets.items():
+        if regular and data["type"] not in ["expansion", "core", "draft_innovation", "commander", "masters"]:
```
should global var that list, since it could expand

ZeldaZach commented on mtgjson4/format.py:
```diff
@@ -0,0 +1,50 @@
+
+        for card in cards:
+            card_formats = set(card.get("legalities").keys())
+            possible_formats &= card_formats
```
I'd like to avoid bitwise ops if possible. Any other way of going about this?

henry739 replied:
Sure there is: It can be expressed as `possible_formats = possible_formats.intersection(card_formats)`

ZeldaZach commented on mtgjson4/outputter.py:
```diff
@@ -349,11 +330,23 @@ def create_and_write_compiled_outputs() -> None:
+    write_to_file(mtgjson4.STANDARD_OUTPUT, __handle_compiling_sets(format_map["standard"], "Standard"))
+
+    # Future.json
+    write_to_file(mtgjson4.FUTURE_OUTPUT, __handle_compiling_sets(format_map["future"], "Future"))
```
I'd like to re-key Future => "Future Standard"

henry739 replied:
Do you want the output to be called "Future Standard"?

ZeldaZach commented on tests/mtgjson4/test_format.py:
```diff
@@ -0,0 +1,68 @@
+from mtgjson4.format import build_format_map
+
+
+NULL_OUTPUT = {
```
Type is required. Also, if this is the same as the list before, you can simply make one list and refer to it as keys and/or values.

ZeldaZach replied:
Err this is a tester, so not sure if we'd want to do what I said above. Can go either way.

ZeldaZach commented on mtgjson4/format.py:
```diff
@@ -0,0 +1,50 @@
+            continue
+
+        possible_formats = set(FORMATS)
+        cards = data.get("cards")
```
don't need a local var, can just put in the for loop

ZeldaZach commented on mtgjson4/format.py:
```diff
@@ -0,0 +1,50 @@
+        cards = data.get("cards")
+
+        for card in cards:
+            card_formats = set(card.get("legalities").keys())
```
I honestly don't know what something like this would report back...

henry739 replied:
I'll add some clarifying comments, but the dictionary keys (format) are only present in the legalities dictionary if the value is != 'Not Legal'.

ZeldaZach reviewed (changes requested):
Good start! Keep up the great work :)

henry739 committed:
address code review comments for format.py

henry739 committed:
moved __init__ constants to more logical section

henry739 committed:
implement per-format cards sets.


improve typing and documentation


add missing type annotations


run black formatter over code

henry739 committed:
create functions for creating AllCards and AllSets subsets.

improved constants in __init__.py and updated the OUTPUTS constant. Reduced the number of formats in the FORMATS constant, and renamed it to the more appropriate SUPPORTED_FORMAT_OUTPUTS

henry739 committed:
move strip_bad_sf_chars to util and remove gamepedia.py

henry739 committed:
fix duplicate import

omfgitsmark reviewed (commented):
Where is `get_modern_sets()` from gamepedia.py now? Or where is it getting the modern legal sets from?

omfgitsmark commented on mtgjson4/outputter.py:
```diff
@@ -4,12 +4,13 @@
 
 import mtgjson4
-from mtgjson4 import util
+from mtgjson4 import util, SUPPORTED_FORMAT_OUTPUTS
```
Is this necessary? `mtgjson4.SUPPORTED_FORMAT_OUTPUTS` should work inline.

henry739 replied:
A matter of style maybe? Do we generally want these constants to be prefixed with the module name?

ZeldaZach replied:
We don't have a style doc for such a thing, but I'm fine with mtgjson4.variable (see my earlier comment about GLOBAL class coming soon :tm: )

ZeldaZach committed:
Merge branch 'master' into issue/enhancement-47-more-formats

ZeldaZach commented on mtgjson4/__init__.py:
```diff
@@ -53,12 +53,34 @@
+STANDARD_OUTPUT: str = "Standard"
+MODERN_OUTPUT: str = "Modern"
 VINTAGE_OUTPUT: str = "Vintage"
+LEGACY_OUTPUT: str = "Legacy"
```
I do want to make a `globals` class in the near future since this is getting a bit large for our init

ZeldaZach commented on mtgjson4/outputter.py:
```diff
@@ -165,6 +166,27 @@ def create_all_cards(files_to_ignore: List[str]) -> Dict[str, Any]:
+    subsets: Dict[str, Dict[str, Any]] = {fmt: {} for fmt in target_formats}
+    for name, data in cards.items():
+        for fmt in target_formats:
+            if data["legalities"].get(fmt) in ("Legal", "Restricted"):
```
Good change to do `if var in set(...)` it seems to be much faster than `if var in list(...)`. I should go back and start profiling this project...

ZeldaZach committed:
Minor cleanups

Signed-off-by: Zach Halpern <ZaHalpern+github@gmail.com>

ZeldaZach reviewed (changes requested):
I've had a chance to test this and there are a few areas of improvement:

- Your method of narrowing down what formats contain what sets works for Standard and Modern, but it looks to be wrong for Pauper, Legacy, Vintage, etc
- Pauper has some issues since cards are legal from Vintage sets, but the method we use to narrow down sets is whole sets at a time (so the uncommons rares etc will knock out the entire set). Same issue as above

- Pauper.json needs work (it's empty)

Maybe I"m overthinking this one?

henry739 committed:
update list of format.json and formatcards.json outputs supported


remove bald return from previous testing


apply black formatting to tests


complete removal of formats

ZeldaZach committed:
Merge remote-tracking branch 'origin/master' into issue/enhancement-47-more-formats

ZeldaZach committed:
Minor cleanup

Signed-off-by: Zach Halpern <ZaHalpern+github@gmail.com>
